Add Macronix Flash support for intel_s1000_crb board.

SavinayDharmappa opened the pull request:
Below are contents of the patch set,

1. Enable SPI Master driver for intel_s1000

2.  Add Macronix Flash support  to generic SPI nor flash drive proposed in following PR #9078 (https://github.com/zephyrproject-rtos/zephyr/pull/9078)

3. Add test application to read/write from macronix flash

vanwinkeljan reviewed (commented):
Please take in to consideration following comment that was placed on #9078:
> Looks to me that there is already more generic SPI NOR driver pull request #7561 (which is based on pr #5456). PR #7561 relies on the SFDP information present in the flash device instead of adding the data hard coded into a struct for every device.
>
>Is till see a use case for this driver if the parameters where passed from dts or kconfig as this would reduce the code size compared to auto detection at run time.

rgundi commented on tests/boards/intel_s1000_crb/src/main.c:
```diff
@@ -19,4 +19,7 @@ K_SEM_DEFINE(thread_sem, 1, 1);
 void main(void)
 {
 	printk("Sample app running on: %s Intel S1000 CRB\n", CONFIG_ARCH);
+	printk("Sample app running on: %s Intel_S1000\n", CONFIG_ARCH);
```
Retain the old printk please.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+
+	printk("SPI flash driver was found!\n");
+
+	printk("\nTest 1: Flash erase\n");
```
Incorrect comment. Actually, no comment is required for the below operation.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+
+	printk("\nTest 1: Flash erase\n");
+
+
```
remove the empty line.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+		printk("   Flash erase succeeded!\n");
+	}
+
+
```
remove the empty line.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+
+	for (i = 2 ; i < TEST_DATA_LEN; i++) {
+		buf[i] = 0x63;
+	}
```
not sure the above for loop is intended. It rewrites buf[0] and buf[1] with 0x63. Probably you need to delete the for loop.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+void test_flash(void)
+{
+	struct device *flash_dev;
+	u8_t buf[TEST_DATA_LEN + 10];
```
any particular reason why buf is TEST_DATA_LEN +10? Please remove +10, if it is not needed.

rgundi commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+
+
+	printf("\nTest 2: Flash write\n");
+	flash_write_protection_set(flash_dev, false);
```
since we are already doing this on line 36, do we need to do it again?

rgundi reviewed (changes requested):
@vanwinkeljan: Agree PR #7561 is more generic. However, I wanted to know if ALL Flash devices have SFDP information present in them. If not, i believe there's place for this driver to address such cases.

Of course, I agree that the ID parameters can be passed through KConfig or DTS to make it more configurable. I think that should be done if we agree to accept this driver.

vanwinkeljan commented:
@rgundi 
Just a side note even if all flash devices would support SFDP, which is possibly not the case, there is still a case for this driver. E.g in case you are making your own HW the chance is very small that you will have different flash devices in a single HW version and as such you do not need the overhead of the SFDP handling.

SavinayDharmappa commented:
@vanwinkeljan I have added support to get ID parameters from Kconfig. Can you review it.

rgundi commented on drivers/flash/Kconfig.nor:
```diff
@@ -0,0 +1,129 @@
+	   - W25Q80
+	   - W25Q16
+	   - W25Q32
+	   - S25FL216K
```
Add MX25UM512

rgundi commented on soc/xtensa/intel_s1000/soc.h:
```diff
@@ -154,6 +154,7 @@ struct soc_dsp_shim_regs {
 #define USB_DW_BASE				0x000A0000
 #define USB_DW_IRQ				0x00000806
 
+#define SPI_DW_IRQ_FLAGS                        0
```
Is this required? If yes, i thought we get it from DTS.

rgundi reviewed (changes requested):
Looks almost good. Some minor comments.
1) The commit message for "intel_s1000: Enable SPI Master driver" has a typo. Please correct intel_s100 to intel_s1000.
2) The commit "board: xtensa: intel_s1000_crb: Add macronix flash support" needs to come before the commit "tests: boards: intel_s1000: Add a test app to read/write from flash".

vanwinkeljan commented on drivers/flash/spi_nor.c:
```diff
@@ -0,0 +1,522 @@
+	CONFIG_SPI_NOR_PAGE_SIZE, CONFIG_SPI_NOR_SECTOR_SIZE,
+	CONFIG_SPI_NOR_SECTORS, CONFIG_SPI_NOR_BLOCK_SIZE,
+	CONFIG_SPI_NOR_BLOCK_ERASE_64K},
+};
```
All this information can passed as a device configuration struct that to ```DEVICE_AND_API_INIT```, in this case there is no need anymore for ```char* name```.

vanwinkeljan commented on drivers/flash/spi_nor.c:
```diff
@@ -0,0 +1,522 @@
+ * @page_size: Size of the pages
+ * @flag: Flags of the flash
+ */
+struct spi_nor_parameters {
```
In case a device configuration is used this struct becomes obsolete.

SavinayDharmappa commented:
@vanwinkeljan I have addressed your comments. Please review it.

SavinayDharmappa commented:
@rgundi  Odering of coommit is changing in git hub only. In my local it is proper. Not sure what is the issue

tbursztyka commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -63,6 +63,12 @@ config HEAP_MEM_POOL_SIZE
 endif # DMA_CAVS
 
+config SPI_0_IRQ_PRI
+	default 3
```
you don't need that here, put it in the relevant fixup file (as you for other settings below), and s1000 has no irq priority support afaik (can't recall) so it would be 0.

tbursztyka commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,70 @@
+	u8_t buf[TEST_DATA_LEN];
+	int i;
+
+	printf("==========================\n");
```
s/printf/printk

But you could also use the logging subsystem which is nicer (since you have macros for info/error/warning and debug)

SavinayDharmappa replied:
done.

tbursztyka commented on dts/bindings/slave/spi,flash.yaml:
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2018, Intel Semiconductor
```
Intel Corporation

tbursztyka commented on dts/bindings/slave/spi,flash.yaml:
```diff
@@ -0,0 +1,25 @@
+#
```
Make a patch just for adding that yaml binding file.

tbursztyka commented on drivers/flash/Kconfig.nor:
```diff
@@ -27,11 +27,6 @@ config SPI_NOR_DRV_NAME
 	   - S25FL216K
 	   - MX25UM512
 
-config SPI_NOR_SPI_NAME
```
Why this removal? (you use that option in your board's kconfig)

vanwinkeljan replied:
@tbursztyka because all SPI related info should come from DTS (see #10894)

tbursztyka replied:
Sure (see my comment about Kconfig.defconfig, it's using Kconfig option while it should get these from DTS), but that change should not be in this patch. It should come in a different one.

vanwinkeljan replied:
Actually I would assume it is part of the first commit as it is introducing a new generic SPI driver

tbursztyka replied:
True

tbursztyka commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -69,6 +69,37 @@ config SPI_0_IRQ_PRI
+	default 0
+
+config SPI_NOR_DRV_NAME
+	default "MX25"
```
that name could come from DTS already, as well as the SPI frequency below. (use fixup to relate these.)

tbursztyka reviewed (dismissed):
DTS integration needs more work

SavinayDharmappa committed:
dts: bindings: slave: Add yaml file to support spi flash

add yaml file to support spi flash

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
board: xtensa: intel_s1000_crb: Add macronix flash support

patch enables necessary configs to add macronix flash
support for intel_s1000_crb board.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa commented:
@tbursztyka I have adrresssed your review comments . Please review it.

tbursztyka commented on soc/xtensa/intel_s1000/dts_fixup.h:
```diff
@@ -46,4 +46,17 @@
+#define CONFIG_SPI_0_BASE_ADDRESS		SNPS_DESIGNWARE_SPI_E000_BASE_ADDRESS
+#define CONFIG_SPI_0_NAME			SNPS_DESIGNWARE_SPI_E000_LABEL
+
+#define CONFIG_SPI_0_IRQ			((SNPS_DESIGNWARE_SPI_E000_IRQ_0 << 16) | \
```
why this complex macro? Can't you just set the right irq number directly in the .dtsi file?

SavinayDharmappa replied:
No it is not possible to get single interrupt number from .dtsi.

tbursztyka commented on include/arch/xtensa/sys_io.h:
```diff
@@ -115,6 +115,10 @@ static ALWAYS_INLINE
 	return ret;
 }
 
+static ALWAYS_INLINE void sys_write16(u16_t data, mem_addr_t addr)
```
Make it in a separate patch.

Also, move it between sys_write/read32 and sys_write/read8, and add sys_read16 as well.

SavinayDharmappa replied:
ok

tbursztyka commented on tests/boards/intel_s1000_crb/src/main.c:
```diff
@@ -19,4 +22,6 @@ K_SEM_DEFINE(thread_sem, 1, 1);
 void main(void)
 {
 	printk("Sample app running on: %s Intel S1000 CRB\n", CONFIG_ARCH);
```
LOG_INF?

SavinayDharmappa replied:
done

tbursztyka commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -66,6 +66,34 @@ endif # DMA_CAVS
+config SPI_NOR_SPI_SLAVE
+	default 0
+
+config SPI_NOR_SPI_FREQ_0
```
that one you can get it also from dts.

SavinayDharmappa replied:
done

tbursztyka commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -31,3 +31,14 @@
+	mx25um512@0 {
+		compatible = "spi,flash";
+		label = "MX25UM512";
+		reg = <0x18>;
```
why 0x18?

SavinayDharmappa replied:
changed it to 0.

vanwinkeljan commented on drivers/flash/Kconfig.nor:
```diff
@@ -0,0 +1,129 @@
+	   - W25Q32
+	   - S25FL216K
+
+config SPI_NOR_SPI_NAME
```
Kconfig still contains SPI interface specific config that should move to DTS:
* SPI_NOR_SPI_FREQ_0
* SPI_NOR_SPI_SLAVE
* SPI_NOR_GPIO_SPI_CS
* SPI_NOR_GPIO_SPI_CS_DRV_NAME
* SPI_NOR_GPIO_SPI_CS_PIN

vanwinkeljan replied:
I think you can already safely remove the GPIO related configs from the KConfig and only keep it in the source file. In this case once you the update for arduino_101 you only need top update the dts and fixup files.

SavinayDharmappa committed:
arch: xtensa: Add I/O functions to read/write.

patch add I/O function to read / write from 16 bit
memory mapped registers

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa commented on drivers/flash/Kconfig.nor:
```diff
@@ -0,0 +1,129 @@
+	default n
+	help
+	This Enables 64K Block erase
+
```
I have done changes and updateded

vanwinkeljan commented on drivers/flash/Kconfig.nor:
```diff
@@ -0,0 +1,105 @@
+endchoice
+
+config SPI_NOR_BLOCK_ERASE_SIZE
+	int "Block size for erasing SPI flash"
```
Just a remark in case you would like that the user can only select values through the choice above remove the text after ```int``` and it will not show up in menu config

SavinayDharmappa replied:
done.

SavinayDharmappa commented on drivers/flash/Kconfig.nor:
```diff
@@ -0,0 +1,105 @@
+
+config SPI_NOR_BLOCK_ERASE_SIZE
+	int "Block size for erasing SPI flash"
+	depends on SPI_NOR_BLOCK_ERASE_64K || SPI_NOR_BLOCK_ERASE_32K
```
I have done changes and updated

vanwinkeljan commented:
@tbursztyka Should we not just close #PR 9078 as this PR  already includes the requested improvements?

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -0,0 +1,453 @@
+	return spi_transceive(driver_data->spi,
+		&driver_data->spi_cfg, &tx_set, &rx_set);
+}
+#define spi_nor_cmd_read(dev, opcode, dest, length) \
```
add an empty line right before

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -0,0 +1,453 @@
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_wait_until_ready(struct device *dev)
```
no need to inline

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -0,0 +1,453 @@
+	return ret;
+}
+
+/**
```
remove all flash API documentation: it's redundant with the generic flash header.

tbursztyka commented:
@vanwinkeljan sure, I added comments then on the first commit that I did on the original PR.

rgundi commented:
@tbursztyka Could you please check and approve?

nashif commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig:
```diff
@@ -52,3 +52,8 @@ CONFIG_UART_NS16550=y
 CONFIG_AUDIO_CODEC=y
 CONFIG_AUDIO_TLV320DAC=y
+
+CONFIG_SPI=y
```
Do not enable those by default in the _defconfig, they should be set by the application.

https://docs.zephyrproject.org/latest/porting/board_porting.html#default-board-configuration

SavinayDharmappa replied:
done.

SavinayDharmappa committed:
drivers: flash: add a generic spi nor flash driver

This driver is inspired from the w25qxxdv SPI NOR flash driver which was
already implementing the CFI (Common Flash Interface) for its purpose.
To handle other NOR flash a flash id table (as Linux do) which contains
the geometry for a few SPI NOR flash based on their JEDEC ID has been
introduced.
We currently support the following flash:
 - W25Q80
 - W25Q16
 - W25Q32
 - S25FL216K
 - MX25UM512

The read and write functions are able to handle more then one page at a
time and return the number of bytes read or write.
Also because every NOR flash expect to disable the write protection
before writing or erasing, the write enable command is now part of the
write and erase functions.

Signed-off-by: Sebastien Bourdelin <sebastien.bourdelin@savoirfairelinux.com>
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

sathishkuttan commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -31,3 +31,14 @@
+		compatible = "spi,flash";
+		label = "MX25UM512";
+		reg = <0x00>;
+		spi-max-frequency = <2000000>;
```
Since this is max frequency, it should be 100MHz. The actual operating frequency (if there is a variable for that) can be 2MHz.

SavinayDharmappa replied:
I don't think we have any such variable in dts.

nashif commented:
rebase still needed

SavinayDharmappa committed:
boards: xtensa: intel_s1000_crb: Enable SPI Master driver

patch enables SPI Master driver on intel_s1000_crb

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
tests: boards: intel_s1000_crb: Add a test app to read/write from flash

patch test SPI flash on intel_s1000_crb board.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa commented:
re-based and resolved merge conflict.
